src/rust/src/lib.rs

```diff
@@ -1,132 +1,224 @@
+    let ncol: usize = columns
+        .clone()
+        .map(|(_id, x)| {
+            if x.is_factor() {
+                x.levels().unwrap().len() - 1
+            } else {
+                1
+            }
+        })
+        .sum();
+
+    // add intercept
+    let ncol = ncol + 1;
+
+    // note that we need one more column for the _intercept_
+    let mut processed_columns_matrix: RMatrix<f64> = RMatrix::new(nrow, ncol);
+    let processed_columns = processed_columns_matrix.as_real_slice_mut().unwrap();
+    let mut column_names: Vec<String> = Vec::with_capacity(ncol);
 
     // Add intercept column
-    processed_columns.push(Array2::ones((nrow, 1)));
+    processed_columns[0..nrow].fill(1.);
+
     column_names.push("intercept".to_string());
 
     // Iterate through columns
-    for (col_name, column) in data.iter() {
-        let (processed_column, mut new_column_names) = match column.rtype() {
+    let mut current_column = 1; // we passed the intercept
+    let mut data_iter = columns;
+    loop {
+        let (col_name, column) = if let Some(next_column) = data_iter.next() {
+            next_column
+        } else {
+            break;
+        };
+
+        match column.rtype() {
             // Note that factors match this first condition
-            Rtype::Integers => process_integer_column(&column, col_name, nrow),
-            Rtype::Doubles => process_double_column(&column, col_name, nrow),
-            Rtype::Strings => process_string_column(&column, col_name, nrow),
-            Rtype::Logicals => process_logical_column(&column, col_name, nrow),
-            _ => return Err(Error::Other(format!("Unsupported column type: {:?}", column.rtype()))),
+            Rtype::Integers if column.is_factor() => {
+                let nlevels = column.levels().unwrap().len() - 1;
+                process_factor_column(
+                    &column,
+                    col_name,
+                    nrow,
+                    &mut column_names,
+                    &mut processed_columns
+                        [(current_column * nrow)..((current_column + nlevels) * nrow)],
+                )
+            }
+            Rtype::Integers => process_integer_column(
+                &column,
+                col_name,
+                &mut column_names,
+                &mut processed_columns[(current_column * nrow)..((current_column + 1) * nrow)],
+            ),
+            Rtype::Doubles => process_double_column(
+                &column,
+                col_name,
+                &mut column_names,
+                &mut processed_columns[(current_column * nrow)..((current_column + 1) * nrow)],
+            ),
+            Rtype::Strings => process_string_column(
+                &column,
+                col_name,
+                nrow,
+                &mut column_names,
+                &mut processed_columns[(current_column * nrow)..],
```

Comment:
if `rayon` is the next step, then this part needs a bit of adjustment. First, I don't think this part is even run anymore, thanks to `columns`.